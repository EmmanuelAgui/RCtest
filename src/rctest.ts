import { Sequence } from "./index";
// @ts-ignore
export let sequence = new Sequence();
console.log(sequence.next());
