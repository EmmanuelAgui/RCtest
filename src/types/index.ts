export enum EXT_TYPE{
    DigitalUser = 'DigitalUser',
    VirtualUser = 'VirtualUser',
    FaxUser = 'VirtualUser',
    AO = 'AO',
    Dept='Dept'
  }

export interface Extension {
    firstName: string
    lastName?: string
    ext?: string
    extType: EXT_TYPE
    fullName?: string // "firstName" + "lastName" + "ext" 
    extTypeNum?:number // map from extType
    [prop:string]:any
}
  // extensions function type
export type ExtensionFun = (extensions: Extension[], sortKey?: keyof Extension, sortType?: 'ASC' | 'DESC') => Extension[];

export interface SaleItem{
    month: number
    date: number
    transationId: string
    salePrice: number
}
  
export interface QuarterItem{
    quarter: number
    transactionNums: number
}
  
export interface QuarterSumItem extends QuarterItem{
    totalPrices: number
}
  
export interface QuarterAverageItem extends QuarterItem{
    averagePrices: number
}