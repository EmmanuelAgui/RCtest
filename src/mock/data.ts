import { EXT_TYPE } from "../types";

// test data
export let myExtensions = [
    { firstName: 'adfsdf', lastName: 'xxddx', ext: 'adfx', extType: EXT_TYPE.AO },
    { firstName: 'rtfgh', lastName: 'sfdf', ext: '', extType: EXT_TYPE.FaxUser },
    { firstName: 'dfdf', lastName: 'dfs', ext: 'adfx', extType: EXT_TYPE.Dept },
    { firstName: 'dfd', lastName: '', ext: 'adfx', extType: EXT_TYPE.Dept },
    { firstName: 'sfwf', lastName: 'ffdfg', ext: '', extType: EXT_TYPE.VirtualUser },
    { firstName: 'reer', lastName: 'dfg', ext: '', extType: EXT_TYPE.DigitalUser },
    { firstName: 'cvfg', lastName: 'hts', ext: 'adfx', extType: EXT_TYPE.AO },
    { firstName: 'bwerew', lastName: '', ext: 'adfx', extType: EXT_TYPE.FaxUser },
    { firstName: 'weewr', lastName: 'derter', ext: 'adfx', extType: EXT_TYPE.Dept }
];
export let mySaleItems = [
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 4,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 5,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 9,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 7,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 5,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    }, {
        month: 11,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    }
];
  