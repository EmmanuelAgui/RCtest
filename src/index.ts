import { ExtensionFun, Extension, EXT_TYPE, SaleItem, QuarterSumItem, QuarterAverageItem } from "./types";
import { setFullNameOfExtensions, sortExtensionsByKey, getSortedExtensions, setExtTypeNumOfExtensions } from "./helpers/extensions";
import { sortSaleItems, sumSalesByQuarter, getQuarterAverageFrom } from "./helpers/sales";
import { myExtensions, mySaleItems } from "./mock/data";
  
/**
    extensions is an Array and each item has such format:
    {firstName: 'xxx', lastName: 'xxx', ext: 'xxx', extType: 'xxx'}
    lastName, ext can be empty, extType can only has "DigitalUser", "VirtualUser","FaxUser","Dept","AO".
**/
  
/**
    Question 1: sort extensions by "firstName" + "lastName" + "ext" ASC
**/
export let sortExtensionsByName: ExtensionFun = (extensions: Extension[]): Extension[] => {
    // extend fullName ("firstName" + "lastName" + "ext") of extension item
    let extensionsWithFullName = setFullNameOfExtensions(extensions);
    // sort extensions by fullName
    let sortedExtensionsWithFullName = sortExtensionsByKey(extensionsWithFullName, 'fullName');
    // get result extensions
    extensions = getSortedExtensions(sortedExtensionsWithFullName);
    return extensions;
}

/**
    Question 2: sort extensions by extType follow these orders ASC
    DigitalUser < VirtualUser < FaxUser < AO < Dept.
**/
export let sortExtensionsByExtType: ExtensionFun = (extensions: Extension[]): Extension[] => {
    // extend extTypeNum of extension item
    let extensionsWithExtTypeNum = setExtTypeNumOfExtensions(extensions);
    // sort extensions by extTypeNum
    let sortedExtensionsWithExtTypeNum = sortExtensionsByKey(extensionsWithExtTypeNum, 'extTypeNum');
    // get result extensions
    extensions = getSortedExtensions(sortedExtensionsWithExtTypeNum);
    return extensions;
}
// // test
// let test = sortExtensionsByName(myExtensions);
// console.log(test);
// test = sortExtensionsByExtType(myExtensions);
// console.log(test);
/**
    saleItems is an Array has each item has such format:
    {
        month: n, //[1-12],
        date: n, //[1-31],
        transationId: "xxx",
        salePrice: number
    }
**/

/**
    Question 3: write a function to calculate and return a list of total sales (sum) for each quarter, expected result like:
    [
        {quarter: 1, totalPrices: xxx, transactionNums: n},
        {....}
    ]
**/
export function sumByQuarter(saleItems: SaleItem[]): QuarterSumItem[]{
    let result: QuarterSumItem[]
    // sort saleItems by month
    saleItems = sortSaleItems(saleItems);
    // calcuate totalPrices and transactionNums of each quarter
    result = sumSalesByQuarter(saleItems);
    return result;
}
/**
    Question 4: write a function to calculate and return a list of average sales for each quarter, expected result like:
    [
        {quarter: 1, averagePrices: xxx, transactionNums: n},
        {....}
    ]
**/

export function averageByQuarter(saleItems: SaleItem[]): QuarterAverageItem[]{
    let result: QuarterAverageItem[];
    // calcuate totalPrices and transactionNums of each quarter
    let sumResult: QuarterSumItem[] = sumByQuarter(saleItems);
    // calcuate averagePrices of each quarter
    result = getQuarterAverageFrom(sumResult);
    return result;
}
//test
let quarterSumItems = sumByQuarter(mySaleItems);
console.log(quarterSumItems);
let quarterAverageItems = averageByQuarter(mySaleItems);
console.log(quarterAverageItems);
  
/**
    Question 5: please create a tool to generate Sequence
    Expected to be used like:
    var sequence1 = new Sequence();
    sequence1.next() --> return 1;
    sequence1.next() --> return 2;
    
    in another module:
    var sequence2 = new Sequence();
    sequence2.next() --> 3;
    sequence2.next() --> 4;
**/
// js Singleton
export const Sequence = (function () {
    // store instance
    let _instance:any=null;

    const Sequence = function () {
    if (_instance) return _instance;// return the instance exist
    // @ts-ignore
    _instance = this;
    // @ts-ignore
    this.init();
    return _instance;
    }

    Sequence.prototype.init = function () {
    _instance.source = 0;
    }
    Sequence.prototype.next = function () {
    return ++_instance.source;
    }

    return Sequence
})()
// @ts-ignore
const s1 = new Sequence();
// @ts-ignore
const s2 = new Sequence();
console.log(s1.next());
console.log(s2.next());
console.log(s1.next());

/**
     Question 6:
    AllKeys: 0-9;
    usedKeys: an array to store all used keys like [2,3,4];
    We want to get an array which contains all the unused keys,in this example it would be: [0,1,5,6,7,8,9]
**/

export function getUnUsedKeys<T>(allKeys:T[], usedKeys:T[]):T[]{
    // @ts-ignore
    let unUsedKeys: T[] = allKeys.filter(key => !usedKeys.includes(key));
    return unUsedKeys;
}
// test
// let allKeys = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
// let usedKeys = [2, 3, 4];
// let unUsedKeys = getUnUsedKeys(allKeys, usedKeys);
// console.log(unUsedKeys);
  