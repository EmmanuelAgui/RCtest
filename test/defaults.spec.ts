import {
    sortExtensionsByName,
    sortExtensionsByExtType,
    sumByQuarter,
    averageByQuarter,
    getUnUsedKeys
} from '../src/index'
import { sequence } from '../src/rctest'
import { myExtensions, mySaleItems } from '../src/mock/data'
import { QuarterSumItem, QuarterAverageItem } from '../src/types'

describe('defaults,main functions', () => {
    test('should sort extensions by "firstName" + "lastName" + "ext" ASC', () => {
        let sortedExtensionsByName = sortExtensionsByName(myExtensions)
        expect(sortedExtensionsByName).toMatchSnapshot()
    })

    test('should sort extensions by extType follow these orders ASC', () => {
        let sortedExtensionsByExtType = sortExtensionsByExtType(myExtensions)
        expect(sortedExtensionsByExtType).toMatchSnapshot()
    })

    test('should calculate and return a list of total sales (sum) for each quarter', () => {
        let quarterSumItems: QuarterSumItem[] = sumByQuarter(mySaleItems)
        expect(quarterSumItems).toMatchSnapshot()
    })

    test('should calculate and return a list of average sales for each quarter', () => {
        let quarterAverageItems: QuarterAverageItem[] = averageByQuarter(mySaleItems)
        expect(quarterAverageItems).toMatchSnapshot()
    })

    test('should be a singleton', () => {
        expect(sequence.next()).toBe(5)
    })

    test('get an array which contains all the unused keys', () => {
        let allKeys = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
        let usedKeys = [2, 3, 4]
        let unUsedKeys = getUnUsedKeys(allKeys, usedKeys)
        expect(unUsedKeys).toEqual([0, 1, 5, 6, 7, 8, 9])
    })
})
