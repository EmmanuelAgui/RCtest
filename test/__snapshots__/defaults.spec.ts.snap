// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults,main functions should calculate and return a list of average sales for each quarter 1`] = `
Array [
  Object {
    "averagePrices": 12,
    "quarter": 1,
    "transactionNums": 7,
  },
  Object {
    "averagePrices": 12,
    "quarter": 2,
    "transactionNums": 2,
  },
  Object {
    "averagePrices": 12,
    "quarter": 3,
    "transactionNums": 2,
  },
  Object {
    "averagePrices": 12,
    "quarter": 4,
    "transactionNums": 1,
  },
]
`;

exports[`defaults,main functions should calculate and return a list of total sales (sum) for each quarter 1`] = `
Array [
  Object {
    "quarter": 1,
    "totalPrices": 84,
    "transactionNums": 7,
  },
  Object {
    "quarter": 2,
    "totalPrices": 24,
    "transactionNums": 2,
  },
  Object {
    "quarter": 3,
    "totalPrices": 24,
    "transactionNums": 2,
  },
  Object {
    "quarter": 4,
    "totalPrices": 12,
    "transactionNums": 1,
  },
]
`;

exports[`defaults,main functions should sort extensions by "firstName" + "lastName" + "ext" ASC 1`] = `
Array [
  Object {
    "ext": "adfx",
    "extType": "AO",
    "firstName": "adfsdf",
    "lastName": "xxddx",
  },
  Object {
    "ext": "adfx",
    "extType": "VirtualUser",
    "firstName": "bwerew",
    "lastName": "",
  },
  Object {
    "ext": "adfx",
    "extType": "AO",
    "firstName": "cvfg",
    "lastName": "hts",
  },
  Object {
    "ext": "adfx",
    "extType": "Dept",
    "firstName": "dfd",
    "lastName": "",
  },
  Object {
    "ext": "adfx",
    "extType": "Dept",
    "firstName": "dfdf",
    "lastName": "dfs",
  },
  Object {
    "ext": "",
    "extType": "DigitalUser",
    "firstName": "reer",
    "lastName": "dfg",
  },
  Object {
    "ext": "",
    "extType": "VirtualUser",
    "firstName": "rtfgh",
    "lastName": "sfdf",
  },
  Object {
    "ext": "",
    "extType": "VirtualUser",
    "firstName": "sfwf",
    "lastName": "ffdfg",
  },
  Object {
    "ext": "adfx",
    "extType": "Dept",
    "firstName": "weewr",
    "lastName": "derter",
  },
]
`;

exports[`defaults,main functions should sort extensions by extType follow these orders ASC 1`] = `
Array [
  Object {
    "ext": "",
    "extType": "DigitalUser",
    "firstName": "reer",
    "lastName": "dfg",
  },
  Object {
    "ext": "",
    "extType": "VirtualUser",
    "firstName": "rtfgh",
    "lastName": "sfdf",
  },
  Object {
    "ext": "",
    "extType": "VirtualUser",
    "firstName": "sfwf",
    "lastName": "ffdfg",
  },
  Object {
    "ext": "adfx",
    "extType": "VirtualUser",
    "firstName": "bwerew",
    "lastName": "",
  },
  Object {
    "ext": "adfx",
    "extType": "AO",
    "firstName": "adfsdf",
    "lastName": "xxddx",
  },
  Object {
    "ext": "adfx",
    "extType": "AO",
    "firstName": "cvfg",
    "lastName": "hts",
  },
  Object {
    "ext": "adfx",
    "extType": "Dept",
    "firstName": "dfdf",
    "lastName": "dfs",
  },
  Object {
    "ext": "adfx",
    "extType": "Dept",
    "firstName": "dfd",
    "lastName": "",
  },
  Object {
    "ext": "adfx",
    "extType": "Dept",
    "firstName": "weewr",
    "lastName": "derter",
  },
]
`;
